src/totp/totp.cpp

```diff
@@ -0,0 +1,124 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+int QTotp::base32_decode(const quint8 *encoded, quint8 *result, int bufSize)
+{
+    int buffer = 0;
+    int bitsLeft = 0;
+    int count = 0;
+
+    for (const quint8 *ptr = encoded; count < bufSize && *ptr; ++ptr) {
+        quint8 ch = *ptr;
+
+        if (ch == ' ' || ch == '\t' || ch == '\r' || ch == '\n' || ch == '-') {
+            continue;
+        }
+
+        buffer <<= 5;
+
+        // Deal with commonly mistyped characters
+        if (ch == '0') {
+            ch = 'O';
+        } else if (ch == '1') {
+            ch = 'L';
+        } else if (ch == '8') {
+            ch = 'B';
+        }
+
+        // Look up one base32 digit
+        if ((ch >= 'A' && ch <= 'Z') || (ch >= 'a' && ch <= 'z')) {
+            ch = (ch & 0x1F) - 1;
+        } else if (ch >= '2' && ch <= '7') {
+            ch -= '2' - 26;
+        } else {
+            return -1;
+        }
+
+        buffer |= ch;
+        bitsLeft += 5;
+
+        if (bitsLeft >= 8) {
+            result[count++] = buffer >> (bitsLeft - 8);
+            bitsLeft -= 8;
+        }
+    }
+
+    if (count < bufSize) {
+        result[count] = '\000';
+    }
+
+    return count;
+}
+
+
+QByteArray QTotp::hmacSha1(QByteArray key, QByteArray baseString)
+{
+    int blockSize = 64;
+    if (key.length() > blockSize) {
+        key = QCryptographicHash::hash(key, QCryptographicHash::Sha1);
+    }
+
+    QByteArray innerPadding(blockSize, char(0x36));
+    QByteArray outerPadding(blockSize, char(0x5c));
+
+    for (int i = 0; i < key.length(); i++) {
+        innerPadding[i] = innerPadding[i] ^ key.at(i);
+        outerPadding[i] = outerPadding[i] ^ key.at(i);
+    }
+
+    QByteArray total = outerPadding;
+    QByteArray part = innerPadding;
+    part.append(baseString);
+    total.append(QCryptographicHash::hash(part, QCryptographicHash::Sha1));
+    QByteArray hashed = QCryptographicHash::hash(total, QCryptographicHash::Sha1);
+    return hashed;
+}
+
+
+QString QTotp::generateTotp(const QByteArray key)
+{
+    quint64 time = QDateTime::currentDateTime().toTime_t();
+    quint64 current = qToBigEndian(time / 30);
+    const unsigned numDigits = 6;
+
+    int secretLen = (key.length() + 7) / 8 * 5;
+    quint8 secret[secretLen];
+    int res = QTotp::base32_decode(reinterpret_cast<const quint8 *>(key.constData()), secret, secretLen);
+    QByteArray hmac = QTotp::hmacSha1(QByteArray(reinterpret_cast<const char *>(secret), res), QByteArray(reinterpret_cast<char*>(&current), sizeof(current)));
```

Comment:
Because it doesn't work that way.

> Returns a byte array containing the string equivalent of the number

Also you need to pay attention to endianness here.